File: funcional/funcional..hs
import System.CPUTime
import Data.List

-- Função para realizar a filtragem recursiva dos dados
filtrarDados :: [[String]] -> Int -> String -> [[String]]
filtrarDados [] _ _ = []
filtrarDados (linha:resto) coluna valor
    | linha !! coluna == valor = filtrarDados resto coluna valor
    | otherwise = linha : filtrarDados resto coluna valor

-- Função para realizar o mapeamento dos dados
mapearDados :: [[String]] -> Int -> (String -> String) -> [[String]]
mapearDados [] _ _ = []
mapearDados (linha:resto) coluna funcao = (map (\(i, valor) -> if i == coluna then funcao valor else valor) (zip [0..] linha)) : mapearDados resto coluna funcao

-- Função para reduzir os dados
reduzirDados :: [[String]] -> Int -> (Float -> Float -> Float) -> Float
reduzirDados [] _ _ = error "Lista vazia"
reduzirDados [linha] coluna _ = read (linha !! coluna)
reduzirDados (linha:resto) coluna funcao = funcao (read (linha !! coluna)) (reduzirDados resto coluna funcao)

-- Função para ordenar os dados
ordenarDados :: [[String]] -> Int -> [[String]]
ordenarDados dados coluna = sortBy (\a b -> compare (a !! coluna) (b !! coluna)) dados

-- Função para carregar os dados de um arquivo CSV
carregarDadosArquivo :: FilePath -> IO [[String]]
carregarDadosArquivo nomeArquivo = do
    conteudo <- readFile nomeArquivo
    let linhas = map (splitOn ',') (lines conteudo)
    return linhas

-- Função auxiliar para dividir uma string em uma lista de strings usando um delimitador
splitOn :: Char -> String -> [String]
splitOn delimiter = foldr f [""] where
    f c l@(x:xs) | c == delimiter = "":l
                 | otherwise = (c:x):xs

-- Exemplo de uso do código
main :: IO ()
main = do
    t1_start <- getCPUTime
    dados <- carregarDadosArquivo "dados.csv" -- UTILIZE QUALQUER CSV que se adeque com o problema.
    t1_stop <- getCPUTime

    let dados_copia = dados

    t2_start <- getCPUTime
    let dados_filtrados = filtrarDados dados_copia 1 "30"
    t2_stop <- getCPUTime

    let dados_filtrados_copia = dados_filtrados

    t3_start <- getCPUTime
    let dados_mapeados = mapearDados dados_filtrados_copia 2 (\x -> show (read x * 1.1))
    t3_stop <- getCPUTime

    let dados_mapeados_copia = dados_mapeados

    t4_start <- getCPUTime
    let media_salarios = reduzirDados dados_mapeados_copia 2 (+) / fromIntegral (length dados_mapeados)
    t4_stop <- getCPUTime

    t5_start <- getCPUTime
    let dados_ordenados = ordenarDados dados_mapeados_copia 0
    t5_stop <- getCPUTime

    putStrLn "Dados Filtrados:"
    print dados_filtrados

    putStrLn "Dados Mapeados:"
    print dados_mapeados

    putStrLn "Média dos Salários:"
    print media_salarios

    putStrLn "Dados Ordenados:"
    print dados_ordenados

    putStrLn $ "Tempo de leitura do arquivo: " ++ show (fromIntegral (t1_stop - t1_start) / (10^12 :: Double)) ++ " segundos"
    putStrLn $ "Tempo de filtragem dos dados: " ++ show (fromIntegral (t2_stop - t2_start) / (10^12 :: Double)) ++ " segundos"
    putStrLn $ "Tempo de mapeamento dos dados: " ++ show (fromIntegral (t3_stop - t3_start) / (10^12 :: Double)) ++ " segundos"
    putStrLn $ "Tempo de redução dos dados: " ++ show (fromIntegral (t4_stop - t4_start) / (10^12 :: Double)) ++ " segundos"
    putStrLn $ "Tempo de ordenação dos dados: " ++ show (fromIntegral (t5_stop - t5_start) / (10^12 :: Double)) ++ " segundos"

File: funcional/funcional.py
from time import process_time
import sys 
sys.setrecursionlimit(10**6) 
# Função para realizar a filtragem recursiva dos dados
def filtrar_dados(dados, coluna, valor):
    if not dados: 
        return []

    if dados[0][coluna] == valor:  
        return filtrar_dados(dados[1:], coluna, valor) 
    
    return [dados[0]] + filtrar_dados(dados[1:], coluna, valor) 

# Função para realizar o mapeamento dos dados
def mapear_dados(dados, coluna, funcao):
    if not dados:
        return []
    dados_copia = dados.copy()  # Criar uma cópia dos dados originais
    dados_copia[0][coluna] = funcao(dados_copia[0][coluna])
    mapear_dados(dados_copia[1:], coluna, funcao)
    return dados_copia

def reduzir_dados(dados, coluna, funcao, index=0, resultado=None):
    if index == len(dados):
        return resultado
    
    valor_atual = float(dados[index][coluna])
    if resultado is None:
        resultado = valor_atual
    else:
        resultado = funcao(resultado, valor_atual)
    
    return reduzir_dados(dados, coluna, funcao, index + 1, resultado)

# Função para ordenar os dados
def ordenar_dados(dados, coluna):
    return sorted(dados, key=lambda x: x[coluna])

# Função auxiliar recursiva para ler as linhas do arquivo
def ler_linhas_arquivo(arquivo):
    linha = arquivo.readline().strip()
    if linha:
        return [linha.split(",")] + ler_linhas_arquivo(arquivo)
    else:
        return []

# Função para carregar os dados de um arquivo CSV
def carregar_dados_arquivo(nome_arquivo):
    with open(nome_arquivo, "r") as arquivo:
        dados = ler_linhas_arquivo(arquivo)
    return dados

# Exemplo de uso do código
t1_start = process_time()
dados = carregar_dados_arquivo("dados.csv") # UTILIZE QUALQUER CSV que se adeque com o problema.
t1_stop = process_time()

dados_copia = dados.copy()

# Filtrar os dados com base na coluna "idade"
t2_start = process_time()
dados_filtrados = filtrar_dados(dados_copia, coluna=1, valor="30")
t2_stop = process_time()
print("Dados Filtrados:")
print(dados_filtrados)

dados_filtrados_copia = dados_filtrados.copy()

# Mapear os dados na coluna "salario" com uma função de aumento
t3_start = process_time()
dados_mapeados = mapear_dados(dados_filtrados_copia, coluna=2, funcao=lambda x: float(x) * 1.1)
t3_stop = process_time()
print("Dados Mapeados:")
print(dados_mapeados)

dados_mapeados_copia = dados_mapeados.copy()

# Reduzir os dados para obter a média dos salários
t4_start = process_time()
media_salarios = reduzir_dados(dados_mapeados_copia, coluna=2, funcao=lambda x, y: x + y) / len(dados_mapeados)
t4_stop = process_time()
print("Média dos Salários:")
print(media_salarios)

# Ordenar os dados pela coluna "nome"
t5_start = process_time()
dados_ordenados = ordenar_dados(dados_mapeados_copia, coluna=0)
t5_stop = process_time()

print("Dados Ordenados:")
print(dados_ordenados)

print("Tempo de leitura do arquivo: ", t1_stop - t1_start)
print("Tempo de filtragem dos dados: ", t2_stop - t2_start)
print("Tempo de mapeamento dos dados: ", t3_stop - t3_start)
print("Tempo de redução dos dados: ", t4_stop - t4_start)
print("Tempo de ordenação dos dados: ", t5_stop - t5_start)
